zcash_client_backend/src/data_api.rs

```diff
@@ -46,6 +51,155 @@ pub enum NullifierQuery {
+    numerator: T,
+    denominator: T,
+}
+
+impl<T> Ratio<T> {
+    /// Constructs a new Ratio from a numerator and a denominator.
+    pub fn new(numerator: T, denominator: T) -> Self {
+        Self {
+            numerator,
+            denominator,
+        }
+    }
+
+    /// Returns the numerator of the ratio.
+    pub fn numerator(&self) -> &T {
+        &self.numerator
+    }
+
+    /// Returns the denominator of the ratio.
+    pub fn denominator(&self) -> &T {
+        &self.denominator
+    }
+}
+
+/// A type representing the potentially-spendable value of unspent outputs in the wallet.
+///
+/// The balances reported using this data structure may overestimate the total spendable value of
+/// the wallet, in the case that the spend of a previously received shielded note has not yet been
+/// detected by the process of scanning the chain. The balances reported using this data structure
+/// can only be certain to be unspent in the case that [`Self::is_synced`] is true, and even in
+/// this circumstance it is possible that a newly created transaction could conflict with a
+/// not-yet-mined transaction in the mempool.
+#[derive(Debug, Clone)]
+pub struct WalletSummary {
+    account_balances: BTreeMap<AccountId, AccountBalance>,
+    chain_tip_height: BlockHeight,
+    fully_scanned_height: BlockHeight,
+    sapling_scan_progress: Option<Ratio<u64>>,
+}
+
+impl WalletSummary {
+    /// Constructs a new [`WalletSummary`] from its constituent parts.
+    pub fn new(
+        account_balances: BTreeMap<AccountId, AccountBalance>,
+        chain_tip_height: BlockHeight,
+        fully_scanned_height: BlockHeight,
+        sapling_scan_progress: Option<Ratio<u64>>,
+    ) -> Self {
+        Self {
+            account_balances,
+            chain_tip_height,
+            fully_scanned_height,
+            sapling_scan_progress,
+        }
+    }
+
+    /// Returns the balances of accounts in the wallet, keyed by account ID.
+    pub fn account_balances(&self) -> &BTreeMap<AccountId, AccountBalance> {
+        &self.account_balances
+    }
+
+    /// Returns the height of the current chain tip.
+    pub fn chain_tip_height(&self) -> BlockHeight {
+        self.chain_tip_height
+    }
+
+    /// Returns the height below which all blocks wallet have been scanned, ignoring blocks below
+    /// the wallet birthday.
+    pub fn fully_scanned_height(&self) -> BlockHeight {
+        self.fully_scanned_height
+    }
+
+    /// Returns the progress of scanning Sapling outputs, in terms of the ratio between notes
+    /// scanned and the total number of notes added to the chain since the wallet birthday.
+    ///
+    /// This ratio should only be used to compute progress percentages, and the numerator and
+    /// denominator should not be treated as authoritative note counts. Returns `None` if the
+    /// wallet is unable to determine the size of the note commitment tree.
+    pub fn sapling_scan_progress(&self) -> Option<Ratio<u64>> {
```

Comment:
Does this need to be Sapling-specific? Once we support Orchard scanning, if we follow the pattern established here (adding an `orchard_scan_progress` method) then the wallet UI would have to somehow merge the progress reports for Sapling and Orchard. It could do that by calculating `(sapling_num + orchard_num) / (sapling_denom + orchard_denom)`, but that would only give a sensible result if we assume that the numerators and denominators represent note counts, which the wallet is not supposed to assume. Therefore we might as well define a single `scan_progress` method and make things simpler.

```suggestion
    /// Returns the progress of scanning outputs, in terms of the ratio between notes scanned
    /// and the total number of notes added to the chain since the wallet birthday.
    ///
    /// This ratio should only be used to compute progress percentages, and the numerator and
    /// denominator should not be treated as authoritative note counts. Returns `None` if the
    /// wallet is unable to determine the size of the note commitment tree(s).
    pub fn scan_progress(&self) -> Option<Ratio<u64>> {
```

Reply:
I disagree, because that is complexity that we know we won't need: neither a user nor a wallet cares about the progress of pools individually. In my opinion the best course when we introduce Orchard will be to rename `sapling_scan_progress` to `scan_progress` and have it do the merging. If we do that now instead then the API is both conceptually simpler, and avoids a future compatibility break.

Reply:
Okay, I've made this change.